WIP: Spell Effects

dmportella opened the pull request:
Adding support for popular spell effects

Currently only doing: **Bless, Enlarge and Reduce**

# New sections called Effects
![image](https://github.com/user-attachments/assets/3f6392a4-298f-450f-9142-663d610386be)

# Bless
![image](https://github.com/user-attachments/assets/2b2a5cbf-19ea-425c-876e-a092f43f92bb)

NOT DONE BLESS ATTACK BONUS YET

# Enlarge
![image](https://github.com/user-attachments/assets/ae8524d9-68ed-4c64-a7dc-dff580cd2729)
![image](https://github.com/user-attachments/assets/684786a9-ec49-4686-9e97-291558881f75)

# Reduce
![image](https://github.com/user-attachments/assets/e30c9866-d8ac-4195-a919-2543a331d34c)
![image](https://github.com/user-attachments/assets/bb71354f-14ad-42c5-bdac-baceb196019c)

dmportella commented:
@kakaroto added all the effects related stuff for bless, and enlarge and reduce, future i will add bane, hunters mark and hex. Hunters mark will allow us to apply the Foe Slayer class feature that changes hunters mark damage to 1d10. and i will probably stop there, and not add any other effect.

kakaroto commented on src/common/roll_renderer.js:
```diff
@@ -503,7 +503,12 @@ class Beyond20RollRenderer {
-        return this.postDescription(request, title, null, {}, null, rolls);
+
+        const roll_info = [];
+        if (request["effects"] !== undefined && request["effects"].length !== 0)
```
I'm thinking of a third party sending a roll and putting `request.effects = null` or even `request.effect = "Bless"` which would cause a crash here.

A safer check would be:
```js
if (Array.isArray(request["effects"]) && request["effects"].length > 0)
```

kakaroto commented on src/dndbeyond/content-scripts/character.js:
```diff
@@ -277,9 +294,84 @@ function rollAbilityOrSavingThrow(paneClass, rollType) {
+    } else if(rollType == "saving-throw" && character.getSetting("effects-bane", false)) {
+        roll_properties["modifier"] += "-1d4";
+        addEffect(roll_properties, "Bane");
+    }
```
I think doing the bless/bane in one place is better. Here you're doing it for ability and saving throws, but then the check happens again in `sendRoll` but only checks for attack and spell-attack.
It's better to remove it from here and expand the list of roll types the check in `sendRoll` applies it to.

kakaroto commented on src/dndbeyond/content-scripts/character.js:
```diff
@@ -277,9 +294,84 @@ function rollAbilityOrSavingThrow(paneClass, rollType) {
+        roll_properties["advantage"] = advantageMapping[roll_properties["advantage"]];
+    }
+    
+    if(key_modifiers.normal_roll) {
```
Unfortunately, you can't override the key modifiers here, because if the hotkey is `N` for normal and `A` for advantage, and you switch it, then when the user releases the letter `N`, the `key_modifiers.normal_roll` will get set to `false` (which it already was) and `key_modifiers.advantage` will not be modified, since the letter `A` was never pressed by the user... and you end up with a "stuck hotkey".

kakaroto reviewed (changes requested):
This is great! Thank you for doing it. I think people will love the feature and I like how you added it to the popup!
Only big concern with this PR is the `key_modifiers` issue, and I'm not sure how to resolve that.... 🤔 humm.. do you even need to set a hotkey to `true`? You're already setting the advantage value, so I think if you just set the values to `false` then it won't be an issue?
Too tired to think clearly, but hopefully I'm on the right track. Let me know what you think!

dmportella commented:
@kakaroto thank you for the review and all the comments, i all the changes suggested and also removed all that modifier stuff out, after testing further and confirming what you said, i moved it to sendroll, which made it a lot simpler. tested with key modifiers and all is working (i no longer change the modifiers) and only changing the advantage variable and all works as intended.

kakaroto commented on src/dndbeyond/content-scripts/character.js:
```diff
@@ -183,6 +190,15 @@ async function rollSkillCheck(paneClass) {
+    } else if(roll_properties.ability == "STR" && character.getSetting("effects-reduce", false)) {
+        adjustRollAndKeyModifiersWithDisadvantage(roll_properties);
+        addEffect(roll_properties, "Reduce");
+    }
```
Forgot to remove the enlarge/reduce effect on the skill checks now that you moved it to `sendRoll`

kakaroto commented on src/dndbeyond/base/utils.js:
```diff
@@ -465,6 +477,22 @@ async function sendRoll(character, rollType, fallback, args) {
+        } else if(req.character.settings["effects-bane"] && req.type === "saving-throw") {
+            req["modifier"] += "-1d4";
+            addEffect(req, "Bane");
+        }
```
The modifier is usually the +x modifier of the ability, so I'd say the bless/bane effect should still be added as a `custom-roll-dice` so you can bundle the check for saving-throw in the previous `if` statement.
```ks
if (req.character.settings["effects-bless"] && 
   (req.type === "saving-throw" || (["attack", "spell-attack"].includes(req.type) && req["to-hit"])) {
```
Actually, I think in the case of attack/spell-attack the custom-roll-dice isn't used if there is no to-hit, so you shouldn't need to check for it, so it could be simplified into
```js
if (req.character.settings["effects-bless"] && ["attack", "spell-attack", "saving-throw"].includes(req.type)) {
```

kakaroto commented on src/dndbeyond/base/utils.js:
```diff
@@ -485,6 +513,14 @@ async function sendRoll(character, rollType, fallback, args) {
+    if((rollType == "saving-throw" || rollType == "ability" || rollType == "skill") && req.ability == "STR" && character.getSetting("effects-enlarge", false)) {
+        adjustRollAndKeyModifiersWithAdvantage(req);
+        addEffect(req, "Enlarge");
+    } else if((rollType == "saving-throw" || rollType == "ability") && req.ability == "STR" && character.getSetting("effects-reduce", false)) {
```
Missing the check for "skill" on the reduce, you only added it on enlarge.
Also, could benefit from using `["saving-throw", "ability", "skill"].includes(rollType)` for cleaner code readability

kakaroto reviewed (changes requested):
Found a few small issues, though nothing major

kakaroto reviewed (approved):
Thanks for the quick fixes!

kakaroto commented on src/extension/popup.js:
```diff
@@ -333,7 +356,8 @@ function populateCharacter(response) {
             $("#beyond20-option-discord-target").addClass("advanced-option");
-        }
+        
+        options.find(".effects-option").hide();}
```
Late night coding and I just noticed this in code, the line seems to have been written before the closing `}` instead of after... so it's inside of the `if` scope above it.
Fixed it, pushing soon!
